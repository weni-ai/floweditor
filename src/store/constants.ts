enum Constants {
  UPDATE_EDITOR_STATE = 'UPDATE_EDITOR_STATE',
  UPDATE_NODE_EDITOR_SETTINGS = 'UPDATE_NODE_EDITOR_SETTINGS',
  UPDATE_TRANSLATING = 'UPDATE_TRANSLATING',
  UPDATE_BASE_LANGUAGE = 'UPDATE_BASE_LANGUAGE',
  UPDATE_SEARCH = 'UPDATE_SEARCH',
  UPDATE_LANGUAGES = 'UPDATE_LANGUAGES',
  UPDATE_LANGUAGE = 'UPDATE_LANGUAGE',
  UPDATE_FETCHING_FLOW = 'FETCHING_FLOW',
  UPDATE_DEFINITION = 'UPDATE_DEFINITION',
  UPDATE_NODE_DRAGGING = 'UPDATE_NODE_DRAGGING',
  UPDATE_FLOWS = 'UPDATE_FLOWS',
  UPDATE_PENDING_CONNECTIONS = 'UPDATE_PENDING_CONNECTIONS',
  REMOVE_PENDING_CONNECTION = 'REMOVE_PENDING_CONNECTION',
  UPDATE_COMPONENTS = 'UPDATE_COMPONENTS',
  UPDATE_CONTACT_FIELDS = 'UPDATE_CONTACT_FIELDS',
  UPDATE_GROUPS = 'UPDATE_GROUPS',
  UPDATE_RESULT_MAP = 'UPDATE_RESULT_MAP',
  UPDATE_ASSET_MAP = 'UPDATE_ASSET_MAP',
  UPDATE_NODES = 'UPDATE_NODES',
  UPDATE_ISSUES = 'UPDATE_ISSUES',
  UPDATE_METADATA = 'UPDATE_METADATA',
  UPDATE_EXITS = 'UPDATE_EXITS',
  UPDATE_NODE_EDITOR_OPEN = 'UPDATE_NODE_EDITOR_OPEN',
  UPDATE_CREATE_NODE_POSITION = 'UPDATE_CREATE_NODE_POSITION',
  UPDATE_DRAG_GROUP = 'UPDATE_DRAG_GROUP',
  UPDATE_TYPE_CONFIG = 'UPDATE_TYPE_CONFIG',
  UPDATE_USER_ADDING_ACTION = 'UPDATE_USER_ADDING_ACTION',
  UPDATE_SEARCH_BAR = 'UPDATE_SEARCH_BAR',
  INCREMENT_SUGGESTED_RESULT_NAME_COUNT = 'INCREMENT_SUGGESTED_RESULT_NAME_COUNT'
}

export default Constants;
