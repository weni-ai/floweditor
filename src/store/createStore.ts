import { applyMiddleware, createStore, Middleware } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension';
import thunk from 'redux-thunk';

import rootReducer from 'store/rootReducer';
import AppState, { initialState } from 'store/state';

const middlewares: Middleware[] = [thunk];

export default (state: AppState = initialState) => {
  const store = createStore(
    rootReducer,
    state,
    composeWithDevTools(applyMiddleware(...middlewares)),
  );

  return store;
};
