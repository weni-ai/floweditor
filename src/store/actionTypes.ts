import { Type } from 'config/interfaces';
import { FlowDefinition, FlowMetadata } from 'flowTypes';
import Constants from 'store/constants';
import { EditorState } from 'store/editor';
import {
  Asset,
  AssetStore,
  ContactFields,
  RenderNodeMap,
  FlowIssueMap,
  Search
} from 'store/flowContext';
import { NodeEditorSettings } from 'store/nodeEditor';

// Redux action generic
interface DuxAction<T extends Constants, P extends { [key: string]: any } = {}> {
  type: T;
  payload?: P;
}

// Payload types
interface EditorStatePayload {
  editorState: EditorState;
}

interface DefinitionPayload {
  definition: FlowDefinition;
}

interface BaseLanguagePayload {
  baseLanguage: Asset;
}

interface SearchPayload {
  search: Search;
}

interface UpdateMetadataPayload {
  metadata: FlowMetadata;
}

interface LanguagesPayload {
  languages: Asset[];
}

interface UpdateFlowsPayload {
  flows: { uuid: string; name: string }[];
}

interface UpdateDependenciesPayload {
  dependencies: FlowDefinition[];
}

interface UpdateAssetMapPayload {
  assets: AssetStore;
}

interface UpdateNodesPayload {
  nodes: RenderNodeMap;
}

interface UpdateIssuesPayload {
  issues: FlowIssueMap;
}

interface UpdateTypeConfigPayload {
  typeConfig: Type;
}

interface UpdateUserAddingActionPayload {
  userAddingAction: boolean;
}

interface UpdateNodeEditorSettingsPayload {
  settings: NodeEditorSettings;
}

interface UpdateContactFieldsPayload {
  contactFields: ContactFields;
}

// Action types
export type UpdateNodeEditorSettings = DuxAction<
  Constants.UPDATE_NODE_EDITOR_SETTINGS,
  UpdateNodeEditorSettingsPayload
>;

export type UpdateEditorState = DuxAction<Constants.UPDATE_EDITOR_STATE, EditorStatePayload>;

export type UpdateBaseLanguageAction = DuxAction<
  Constants.UPDATE_BASE_LANGUAGE,
  BaseLanguagePayload
>;

export type UpdateSearchAction = DuxAction<Constants.UPDATE_SEARCH, SearchPayload>;

export type UpdateLanguagesAction = DuxAction<Constants.UPDATE_LANGUAGES, LanguagesPayload>;

export type UpdateDefinitionAction = DuxAction<Constants.UPDATE_DEFINITION, DefinitionPayload>;

export type UpdateFlowsAction = DuxAction<Constants.UPDATE_FLOWS, UpdateFlowsPayload>;

export type UpdateMetadataAction = DuxAction<Constants.UPDATE_METADATA, UpdateMetadataPayload>;

export type UpdateAssetsAction = DuxAction<Constants.UPDATE_ASSET_MAP, UpdateAssetMapPayload>;

export type IncrementSuggestedResultNameCountAction = DuxAction<
  Constants.INCREMENT_SUGGESTED_RESULT_NAME_COUNT
>;

export type UpdateNodesAction = DuxAction<Constants.UPDATE_NODES, UpdateNodesPayload>;

export type UpdateIssuesAction = DuxAction<Constants.UPDATE_ISSUES, UpdateIssuesPayload>;

export type UpdateTypeConfigAction = DuxAction<
  Constants.UPDATE_TYPE_CONFIG,
  UpdateTypeConfigPayload
>;

export type UpdateUserAddingActionAction = DuxAction<
  Constants.UPDATE_USER_ADDING_ACTION,
  UpdateUserAddingActionPayload
>;

export type UpdateContactFieldsAction = DuxAction<
  Constants.UPDATE_CONTACT_FIELDS,
  UpdateContactFieldsPayload
>;

export type UpdateTypeConfig = (typeConfig: Type) => UpdateTypeConfigAction;

export type UpdateUserAddingAction = (userAddingAction: boolean) => UpdateUserAddingActionAction;

export type UpdateBaseLanguage = (baseLanguage: Asset) => UpdateBaseLanguageAction;

export type UpdateSearch = (search: Search) => UpdateSearch;

export type IncrementSuggestedResultNameCount = () => IncrementSuggestedResultNameCountAction;

type ActionTypes =
  | UpdateEditorState
  | UpdateNodeEditorSettings
  | UpdateDefinitionAction
  | UpdateFlowsAction
  | UpdateMetadataAction
  | UpdateAssetsAction
  | IncrementSuggestedResultNameCountAction
  | UpdateNodesAction
  | UpdateIssuesAction
  | UpdateTypeConfigAction
  | UpdateUserAddingActionAction
  | UpdateBaseLanguageAction
  | UpdateLanguagesAction
  | UpdateContactFieldsAction
  | UpdateSearchAction;

export default ActionTypes;
