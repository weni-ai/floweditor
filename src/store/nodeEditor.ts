import { hasErrors } from 'components/flow/actions/helpers';
import {
  SelectOption,
  UnnnicSelectOption,
} from 'components/form/select/SelectElement';
import { Type } from 'config/interfaces';
import { Scheme } from 'config/typeConfigs';
import { AnyAction } from 'flowTypes';
import mutate from 'immutability-helper';
import { combineReducers } from 'redux';
import { LocalizedObject } from 'services/Localization';
import ActionTypes, {
  UpdateNodeEditorSettings,
  UpdateTypeConfigAction,
  UpdateUserAddingActionAction,
} from 'store/actionTypes';
import Constants from 'store/constants';
import { Asset, RenderNode } from 'store/flowContext';

// tslint:disable:no-shadowed-variable
export interface ValidationFailure {
  message: string;
  field?: string;
}

export interface FormEntry<T = any> {
  value: T;
  validationFailures?: ValidationFailure[];
}

export interface UnnnicSelectOptionEntry<T = any> extends FormEntry {
  value: UnnnicSelectOption<T>;
}

export interface SelectOptionEntry<T = any> extends FormEntry {
  value: SelectOption<T>;
}

export interface NumberEntry extends FormEntry {
  value: number;
}

export interface StringEntry extends FormEntry {
  value: string;
}

export interface SchemeEntry extends FormEntry {
  value: Scheme;
}

export interface StringArrayEntry extends FormEntry {
  value: string[];
}

export interface AssetEntry extends FormEntry {
  value: Asset;
}

export interface AssetArrayEntry extends FormEntry {
  value: Asset[] | null;
}

export interface SelectOptionArrayEntry extends FormEntry {
  value: SelectOption[] | null;
}

export const mergeForm = (
  form: FormState,
  toMerge: Partial<FormState>,
  toRemove: any[] = [],
): FormState => {
  // TODO: deal with explicit array setting
  let updated = form || {};
  // we auto update array items with uuids
  for (const key of Object.keys(toMerge)) {
    const entry: any = (toMerge as any)[key];
    if (Array.isArray(entry)) {
      for (const item of entry) {
        // we support objects with uuids or FormEntry's with uuids
        const isEntry =
          item.hasOwnProperty('value') && typeof item.value === 'object';

        if ((isEntry && item.value.uuid) || item.uuid) {
          const existingIdx = (form as any)[key].findIndex((existing: any) => {
            if (isEntry) {
              return existing.value.uuid === item.value.uuid;
            } else {
              return existing.uuid === item.uuid;
            }
          });

          if (existingIdx > -1) {
            // we found a match, merge us in
            updated = mutate(updated, {
              [key]: { $merge: { [existingIdx]: item } },
            }) as FormState;
          } else {
            // couldn't find it, lets push it on
            updated = mutate(updated, {
              [key]: { $push: [item] },
            }) as FormState;
          }
        }
      }

      // remove it from our merge
      delete (toMerge as any)[key];
    }
  }

  // removals can be items in an array
  for (const remove of toRemove.filter(
    (item: any) => typeof item === 'object',
  )) {
    for (const key of Object.keys(remove)) {
      const entry: any = remove[key];
      if (Array.isArray(entry)) {
        for (const item of entry) {
          // we support objects with uuids or FormEntry's with uuids
          const isEntry =
            item.hasOwnProperty('value') && typeof item.value === 'object';
          if ((isEntry && item.value.uuid) || item.uuid) {
            updated = mutate(updated, {
              [key]: (items: any) =>
                items.filter((existing: any) => {
                  if (isEntry) {
                    return existing.value.uuid !== item.value.uuid;
                  } else {
                    return existing.uuid !== item.uuid;
                  }
                }),
            });
          }
        }
      }
    }
  }

  const removeKeys = toRemove.filter((item: any) => typeof item === 'string');
  updated = mutate(updated, {
    $merge: toMerge,
    $unset: removeKeys,
  }) as FormState;

  let valid = true;
  for (const key of Object.keys(form)) {
    const entry: any = (updated as any)[key];
    if (Array.isArray(entry)) {
      for (const item of entry) {
        if (hasErrors(item)) {
          valid = false;
          break;
        }
      }
    } else if (entry && typeof entry === 'object') {
      if (hasErrors(entry)) {
        valid = false;
        break;
      }
    }
  }

  return mutate(updated, { $merge: { valid } }) as FormState;
};

export interface FormState {
  validationFailures?: ValidationFailure[];
  valid: boolean;
}

export interface NodeEditorSettings {
  originalNode: RenderNode;
  showAdvanced?: boolean;
  originalAction?: AnyAction;
  localizations?: LocalizedObject[];
}

export interface NodeEditor {
  typeConfig: Type | null;
  userAddingAction: boolean;
  settings: NodeEditorSettings | null;
}

// Initial state
export const initialState: NodeEditor = {
  typeConfig: null,
  userAddingAction: false,
  settings: null,
};

// Action Creators
export const updateTypeConfig = (typeConfig: Type): UpdateTypeConfigAction => ({
  type: Constants.UPDATE_TYPE_CONFIG,
  payload: {
    typeConfig,
  },
});

export const updateNodeEditorSettings = (
  settings: NodeEditorSettings,
): UpdateNodeEditorSettings => ({
  type: Constants.UPDATE_NODE_EDITOR_SETTINGS,
  payload: {
    settings,
  },
});

export const updateUserAddingAction = (
  userAddingAction: boolean,
): UpdateUserAddingActionAction => ({
  type: Constants.UPDATE_USER_ADDING_ACTION,
  payload: {
    userAddingAction,
  },
});

// Reducers
export const typeConfig = (
  state: Type | null = initialState.typeConfig,
  action: ActionTypes,
) => {
  switch (action.type) {
    case Constants.UPDATE_TYPE_CONFIG:
      return action.payload!.typeConfig;
    default:
      return state;
  }
};

export const userAddingAction = (
  state: boolean = initialState.userAddingAction,
  action: ActionTypes,
) => {
  switch (action.type) {
    case Constants.UPDATE_USER_ADDING_ACTION:
      return action.payload!.userAddingAction;
    default:
      return state;
  }
};

export const settings = (
  state: NodeEditorSettings | null = initialState.settings,
  action: ActionTypes,
) => {
  switch (action.type) {
    case Constants.UPDATE_NODE_EDITOR_SETTINGS:
      return action.payload!.settings;
    default:
      return state;
  }
};

// Root reducer
export default combineReducers({
  typeConfig,
  userAddingAction,
  settings,
});
