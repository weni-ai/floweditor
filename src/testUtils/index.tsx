// TODO: Remove use of Function
// tslint:disable:ban-types
import { ConfigProviderContext, fakePropType } from 'config/ConfigProvider';
import { mount, ReactWrapper, shallow, ShallowWrapper } from 'enzyme';
import { FlowDefinition, FlowEditorConfig } from 'flowTypes';
import mutate, { Query } from 'immutability-helper';
import * as React from 'react';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import createStore from 'store/createStore';
import { RenderNodeMap } from 'store/flowContext';
import { getFlowComponents } from 'store/helpers';
import AppState, { initialState } from 'store/state';
import config from 'test/config';
import * as matchers from 'testUtils/matchers';
import { merge, set } from 'utils';

// we need to use require syntax to bust implicit any
const boring: FlowDefinition = require('test/flows/boring.json');

// force our matchers to be read in
const match = matchers;

export interface Resp {
  assets: { [key: string]: any }[];
}

export interface QueryString {
  [key: string]: string;
}

export const contextTypes: { [key: string]: Function } = {
  config: fakePropType,
  store: fakePropType,
  assetService: fakePropType,
};

export const baseState: AppState = mutate(initialState, {
  flowContext: merge({
    definition: require('test/flows/colors.json') as FlowDefinition,
  }),
  editorState: merge({
    language: { id: 'eng', name: 'English' },
  }),
});

const flowEditorConfig: FlowEditorConfig = config;

export const configProviderContext: ConfigProviderContext = {
  config: flowEditorConfig,
};

export const setMock = (
  implementation?: (...args: any[]) => any,
): Query<jest.Mock> => set(jest.fn(implementation));

/**
 * Compose setup method for component tests
 */
export const composeSetup = <P extends {}>(
  Component: React.ComponentClass | React.SFC,
  baseProps: P = {} as any,
  baseDuxState: AppState | Partial<AppState> = baseState,
  baseContext: ConfigProviderContext = configProviderContext,
) => (
  shallowRender: boolean = true,
  propOverrides: Query<P | Partial<P>> = {},
  duxStateOverrides: Query<AppState | Partial<AppState>> = {},
  contextOverrides: Query<
    ConfigProviderContext | Partial<ConfigProviderContext>
  > = {},
  childContextTypeOverrides: { [key: string]: Function } = {},
  children: (JSX.Element | React.ComponentClass | React.SFC)[] = [],
) => {
  const props = mutate(baseProps, propOverrides);
  const store = createStore(mutate(
    baseDuxState,
    duxStateOverrides,
  ) as AppState);
  let context = mutate(baseContext, merge({ store }));

  if (Object.keys(duxStateOverrides).length > 0) {
    context = mutate(context, duxStateOverrides);
  }

  if (Object.keys(contextOverrides).length > 0) {
    context = mutate(context, contextOverrides as ConfigProviderContext);
  }

  const childContextTypes: { [contextProp: string]: Function } = mutate(
    contextTypes,
    childContextTypeOverrides,
  );
  // tslint:disable-next-line:ban-types
  const wrapper = (shallowRender ? (shallow as Function) : (mount as Function))(
    <Component {...props}>{children}</Component>,
    {
      context,
      childContextTypes,
    },
  );

  return {
    wrapper,
    props,
    context,
    instance: wrapper.instance(),
  };
};

export const composeSpy = (obj: Record<string, any> | React.ComponentClass) => (
  instanceMethod: string,
) =>
  jest.spyOn(
    (obj as React.ComponentClass).prototype || obj,
    instanceMethod as any,
  );

/**
 * Wait for promises in queue to resolve
 */
export const flushPromises = () =>
  new Promise((resolve: any) => setImmediate(resolve));

/**
 * Restore spy mocks (distinct from mocks created w/ jest.fn()).
 * Use to declaratively restore multiple spy mocks,
 * otherwise just call .mockRestore() on lone spy.
 */
export const restoreSpies = (...spies: jest.SpyInstance<any>[]) =>
  spies.forEach((spy: any) => spy.mockRestore());

// To-do: type this method's output, can pass it prop generic ingested by getComponentTestUtils
/**
 * NOTE: borrowed from EventBrite: https://github.com/eventbrite/javascript/blob/master/react/testing.md#finding-nodes
 * Finds all instances of components in the rendered `componentWrapper` that are DOM components
 * with the `data-spec` attribute matching `name`.
 * @param {ReactWrapper|ShallowWrapper} componentWrapper - Rendered componentWrapper (result of mount, shallow, or render)
 * @param  {string} snacName - Name of `data-spec` attribute value to find
 * @param {string|Function} (Optional) typeFilter - (Optional) Expected type of the wrappers (defaults to all HTML tags)
 * @returns {ReactWrapper|ReactWrapper[]|ShallowWrapper|ShallowWrapper[]} Matching DOM components
 */
export const getSpecWrapper = (
  componentWrapper: ReactWrapper<{}, {}> | ShallowWrapper<{}, {}>,
  specName: string,
  attributeName: string = 'data-spec',
): any => componentWrapper.find(`[${attributeName}="${specName}"]`);

export const composeDuxState = (
  query: Query<AppState | Partial<AppState>> = {},
  duxState = baseState,
) => mutate(duxState, query);

export const composeComponentTestUtils = <P extends {}>(
  Component: React.ComponentClass | React.SFC | any,
  baseProps: P = {} as any,
  baseDuxState: AppState | Partial<AppState> = baseState,
  baseContext: ConfigProviderContext = configProviderContext,
) => ({
  setup: composeSetup<P>(Component, baseProps, baseDuxState, baseContext),
  spyOn: composeSpy(Component),
});

export const prepMockDuxState = (): {
  testNodes: RenderNodeMap;
  mockDuxState: AppState;
} => {
  const testNodes = getFlowComponents(boring).renderNodeMap;

  return {
    testNodes,
    mockDuxState: {
      ...initialState,
      flowContext: {
        ...initialState.flowContext,
        definition: boring,
        nodes: testNodes,
      },
    },
  };
};

export const createMockStore: Function = configureStore([thunk]);

export const mock = <T extends {}, K extends keyof T>(
  object: T,
  property: K,
  value: T[K],
) => {
  Object.defineProperty(object, property, { get: () => value });
};
