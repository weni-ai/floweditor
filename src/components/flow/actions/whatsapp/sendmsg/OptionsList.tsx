import { react as bindCallbacks } from 'auto-bind';
import * as React from 'react';

import styles from 'components/flow/actions/whatsapp/sendmsg/OptionsList.module.scss';
import i18n from 'config/i18n';
import TextInputElement, {
  TextInputSizes,
} from 'components/form/textinput/TextInputElement';
import ContentCollapse from 'components/contentcollapse/ContentCollapse';

import { applyVueInReact } from 'veaury';
// @ts-ignore
import Unnnic from '@weni/unnnic-system';
import {
  SortEnd,
  SortableContainer,
  SortableElement,
  arrayMove,
} from 'react-sortable-hoc';
import { WhatsAppListItem } from 'components/flow/actions/whatsapp/sendmsg/SendWhatsAppMsgForm';
import { MAX_LIST_ITEMS_COUNT } from 'components/flow/actions/whatsapp/sendmsg/constants';
import { CSSTransition, TransitionGroup } from 'react-transition-group';
import update from 'immutability-helper';
import { FormEntry, ValidationFailure } from 'store/nodeEditor';

const UnnnicButton = applyVueInReact(Unnnic.unnnicButton, {
  vue: {
    componentWrap: 'div',
    slotWrap: 'div',
    componentWrapAttrs: {
      style: {
        all: '',
      },
    },
  },
});

export interface OptionsListProps {
  options: FormEntry<WhatsAppListItem[]>;
  onOptionsUpdated(options: WhatsAppListItem[]): void;
  onOptionRemoval(option: WhatsAppListItem): void;
}

export function hasEmptyListItem(listItems: WhatsAppListItem[]): boolean {
  return (
    listItems.find(
      (item: WhatsAppListItem) => item.title.trim().length === 0,
    ) != null
  );
}

const SortableListItem = SortableElement(({ value: row, index }: any) => {
  const listItem = row.item as WhatsAppListItem;
  return (
    <ContentCollapse
      title={`${i18n.t('forms.list_option', 'Option')} ${index +
        1}/${MAX_LIST_ITEMS_COUNT}`}
      open={
        (index === 0 && row.length === 1) || listItem.title.trim().length === 0
      }
      titleIcon="check_circle"
      titleIconScheme={
        listItem.title.trim().length > 0 ? 'weni-600' : 'neutral-soft'
      }
    >
      <div className={styles.list_item_wrapper}>
        <TextInputElement
          name={i18n.t('forms.title', 'Title')}
          placeholder={i18n.t('forms.ex_orange', 'Ex: Orange')}
          size={TextInputSizes.sm}
          onChange={(value, name) =>
            row.list.handleListItemTitleUpdate(value, name, index)
          }
          entry={{ value: listItem.title }}
          autocomplete={true}
          showLabel={true}
          maxLength={24}
        />

        <TextInputElement
          name={i18n.t('forms.description_optional', 'Description (Optional)')}
          placeholder={i18n.t(
            'forms.ex_citrus_and_sweet',
            'Ex: Citrus and sweet',
          )}
          size={TextInputSizes.sm}
          onChange={(value, name) =>
            row.list.handleListItemDescriptionUpdate(value, name, index)
          }
          entry={{ value: listItem.description }}
          autocomplete={true}
          showLabel={true}
          maxLength={72}
        />

        <UnnnicButton
          className={styles.list_item_remove}
          data-testid="Remove"
          iconLeft="do_not_disturb_on"
          text={i18n.t('forms.remove', 'Remove')}
          size="small"
          type="tertiary"
          onClick={() => row.list.handleListItemRemoval(listItem)}
          disabled={
            index === row.length - 1 && listItem.title.trim().length === 0
          }
        />
      </div>
    </ContentCollapse>
  );
});

export default class OptionsList extends React.Component<OptionsListProps> {
  constructor(props: OptionsListProps) {
    super(props);

    bindCallbacks(this, {
      include: [/^on/, /^handle/],
    });
  }

  private handleListItemTitleUpdate(
    value: string,
    name: string,
    index: number,
  ): void {
    // should update name but keep the description intact
    const listItems = update(this.props.options.value, {
      [index]: {
        $merge: {
          title: value,
        },
      },
    }) as WhatsAppListItem[];

    this.props.onOptionsUpdated(listItems);
  }

  private handleListItemDescriptionUpdate(
    value: string,
    name: string,
    index: number,
  ): void {
    //should update description but keep the name intact
    const listItems = update(this.props.options.value, {
      [index]: {
        $merge: {
          description: value,
        },
      },
    }) as WhatsAppListItem[];

    return this.props.onOptionsUpdated(listItems);
  }

  private handleListItemRemoval(item: WhatsAppListItem): void {
    const listItems = this.props.options.value;
    if (
      listItems.indexOf(item) === listItems.length - 1 &&
      hasEmptyListItem(listItems)
    ) {
      return;
    }

    this.props.onOptionRemoval(item);
  }

  /* istanbul ignore next -- @preserve */
  private handleListItemsSortEnd({ oldIndex, newIndex }: SortEnd): void {
    const options = this.props.options.value;
    const listItems = arrayMove(
      options,
      oldIndex,
      newIndex === options.length - 1 ? newIndex - 1 : newIndex,
    ) as WhatsAppListItem[];

    this.props.onOptionsUpdated(listItems);
  }

  private sortableListOptions = SortableContainer(
    ({ items }: { items: WhatsAppListItem[]; shouldCancelStart: any }) => {
      return (
        <TransitionGroup className={styles.list_items}>
          {items.map((value: WhatsAppListItem, index: number) => {
            return (
              <CSSTransition
                key={value.uuid}
                timeout={300}
                classNames={{
                  enter: styles.list_item_enter,
                  enterActive: styles.list_item_enter_active,
                  exit: styles.list_item_exit,
                  exitActive: styles.list_item_exit_active,
                }}
              >
                <SortableListItem
                  key={value.uuid}
                  index={index}
                  value={{ item: value, list: this, length: items.length }}
                  disabled={index === this.props.options.value.length - 1}
                  shouldCancelStart={
                    /* istanbul ignore next -- @preserve */
                    (e: any) => {
                      console.log(e);
                      return true;
                    }
                  }
                />
              </CSSTransition>
            );
          })}
        </TransitionGroup>
      );
    },
  );

  public render(): JSX.Element {
    const hasError =
      this.props.options.validationFailures &&
      this.props.options.validationFailures.length > 0;

    return (
      <div className={styles.list_wrapper}>
        <div>
          <span className={styles.options_label}>
            {i18n.t('forms.options_label', 'Add up to 10 options')}
          </span>

          <this.sortableListOptions
            items={this.props.options.value}
            onSortEnd={this.handleListItemsSortEnd}
            helperClass={styles.item_z_index}
            shouldCancelStart={
              /* istanbul ignore next -- @preserve */
              (e: React.MouseEvent<HTMLDivElement>) => {
                if (e.target.constructor.name === 'SVGSVGElement') {
                  return !(e.target as any).dataset.draggable;
                }

                if (!(e.target instanceof HTMLElement)) {
                  return true;
                }
                return !e.target.dataset.draggable;
              }
            }
            lockAxis="y"
          />

          {hasError ? (
            <div data-testid="Error message" className={styles.error_message}>
              {this.props.options.validationFailures.map(
                (error: ValidationFailure) => error.message,
              )}
            </div>
          ) : null}
        </div>
      </div>
    );
  }
}
