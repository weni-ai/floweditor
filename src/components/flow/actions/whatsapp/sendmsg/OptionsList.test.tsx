import { mock } from 'testUtils';
import OptionsList from './OptionsList';
import * as utils from 'utils';
import * as React from 'react';
import { act, fireUnnnicInputChangeText, render, wait } from 'test/utils';
import { ValidationFailure } from 'store/nodeEditor';

mock(utils, 'createUUID', utils.seededUUIDs());

function getProps() {
  return {
    buttonText: {
      value: 'button text',
    },
    options: {
      value: [
        { uuid: '1', title: 'title 1', description: 'description 1' },
        { uuid: '2', title: 'title 2', description: 'description 2' },
      ],
      validationFailures: [] as ValidationFailure[],
    },
    onOptionsUpdated: jest.fn(),
    onOptionRemoval: jest.fn(),
    onButtonTextUpdated: jest.fn(),
  };
}

describe(OptionsList.name, () => {
  it('should render', async () => {
    const props = getProps();
    const { baseElement } = render(<OptionsList {...props} />);

    await wait();

    expect(baseElement).toMatchSnapshot();
  });

  it('should update list item title', async () => {
    const props = getProps();
    const { getByTestId, getAllByTestId } = render(<OptionsList {...props} />);

    await wait();

    // expand collapse
    const expandCollapseButton = getByTestId('Option 2/10');
    await act(async () => {
      expandCollapseButton.click();
    });

    const listItemTitleInput = getByTestId('Title');
    await act(async () => {
      fireUnnnicInputChangeText(listItemTitleInput, 'new title 2');
    });

    expect(props.onOptionsUpdated).toHaveBeenCalledWith([
      { uuid: '1', title: 'title 1', description: 'description 1' },
      { uuid: '2', title: 'new title 2', description: 'description 2' },
    ]);
  });

  it('should update list item description', async () => {
    const props = getProps();
    const { getByTestId, getAllByTestId } = render(<OptionsList {...props} />);

    await wait();

    // expand collapse
    const expandCollapseButton = getByTestId('Option 2/10');
    await act(async () => {
      expandCollapseButton.click();
    });

    const listItemDescriptionInput = getByTestId('Description (Optional)');
    await act(async () => {
      fireUnnnicInputChangeText(listItemDescriptionInput, 'new description 2');
    });

    expect(props.onOptionsUpdated).toHaveBeenCalledWith([
      { uuid: '1', title: 'title 1', description: 'description 1' },
      { uuid: '2', title: 'title 2', description: 'new description 2' },
    ]);
  });

  it('should remove list item', async () => {
    const props = getProps();
    const { getByTestId, getAllByTestId } = render(<OptionsList {...props} />);

    await wait();

    // expand collapse
    const expandCollapseButton = getByTestId('Option 2/10');
    await act(async () => {
      expandCollapseButton.click();
    });

    const listItemRemoveButton = getByTestId('Remove');
    await act(async () => {
      listItemRemoveButton.click();
    });

    expect(props.onOptionRemoval).toHaveBeenCalledWith({
      uuid: '2',
      title: 'title 2',
      description: 'description 2',
    });
  });

  it('should not remove last list item if it is an empty one', async () => {
    const props = getProps();
    props.options.value = [{ uuid: '1', title: '', description: '' }];
    const { getByTestId, getAllByTestId } = render(<OptionsList {...props} />);

    await wait();

    const listItemRemoveButton = getByTestId('Remove');
    await act(async () => {
      listItemRemoveButton.click();
    });

    expect(props.onOptionRemoval).not.toHaveBeenCalled();
  });

  it('should show error message', () => {
    const props = getProps();
    props.options.validationFailures = [{ message: 'error message text' }];
    const { getByTestId } = render(<OptionsList {...props} />);

    const errorMessage = getByTestId('Error message');

    expect(errorMessage).toBeInTheDocument();
  });
});
