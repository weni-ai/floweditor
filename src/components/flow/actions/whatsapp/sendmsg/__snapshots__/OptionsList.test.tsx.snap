// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OptionsList should render 1`] = `
<body>
  <div>
    <div
      class="list_wrapper"
    >
      <div>
        <span
          class="options_label"
        >
          Add up to 10 options
        </span>
        <div
          class="list_items"
        >
          <div
            class="collapse "
          >
            <div
              class="collapse_header"
              data-testid="Option 1/10"
            >
              <div
                data-use-vue-component-wrap=""
                style="display: flex;"
              >
                <span
                  class="material-symbols-rounded unnnic-icon-scheme--weni-600 unnnic-icon-size--md material-symbols-rounded--filled"
                  data-passed-props="[object Object]"
                  data-v-4ce83152=""
                >
                   check_circle 
                </span>
              </div>
              <span>
                Option 1/10
              </span>
              <div
                class="collapse_icon"
              >
                <div
                  data-use-vue-component-wrap=""
                  style="display: flex;"
                >
                  <span
                    class="material-symbols-rounded unnnic-icon-scheme--neutral-darkest unnnic-icon-size--md"
                    data-passed-props="[object Object]"
                    data-v-4ce83152=""
                  >
                     expand_more 
                  </span>
                </div>
              </div>
            </div>
          </div>
          <div
            class="collapse "
          >
            <div
              class="collapse_header"
              data-testid="Option 2/10"
            >
              <div
                data-use-vue-component-wrap=""
                style="display: flex;"
              >
                <span
                  class="material-symbols-rounded unnnic-icon-scheme--weni-600 unnnic-icon-size--md material-symbols-rounded--filled"
                  data-passed-props="[object Object]"
                  data-v-4ce83152=""
                >
                   check_circle 
                </span>
              </div>
              <span>
                Option 2/10
              </span>
              <div
                class="collapse_icon"
              >
                <div
                  data-use-vue-component-wrap=""
                  style="display: flex;"
                >
                  <span
                    class="material-symbols-rounded unnnic-icon-scheme--neutral-darkest unnnic-icon-size--md"
                    data-passed-props="[object Object]"
                    data-v-4ce83152=""
                  >
                     expand_more 
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
